### PR TITLE
Enhancement: reuse the current content regex

### PR DESCRIPTION
**Pull request by azaozz**
Based on https://github.com/WordPress/wp-lazy-loading/pull/20. This is another "patch" for https://github.com/WordPress/wp-lazy-loading/issues/2 with some changes and enhancements.

Additions/changes compared to pull/20:
- Add `loading` attribute to all img tags in the content, not just tags that have `wp-image-####` class names.
- Tweak/fix functions and vars names to make them more self-documenting.
- Introduce `wp_image_tag_add_srcset_and_sizes_attr()` to go with `wp_image_tag_add_loading_attr()`. It makes it possible to bypass adding of srcset and sizes attributes on a per-image basis, before doing anything else. This also prepares it for adding filtering of `iframe` in the future (`wp_iframe_tag_add_loading_attr()`, etc.).
- Removes the limitation from `wp_filter_content_tags()` to only add srcset and sizes when `$context` is set to `the_content` (i.e. only when run as callback to `the_content` filter). This makes the function useful for filtering any HTML string and adding all supported img tag attributes.

### CODE REVIEW

**Comment by azaozz** [comment]
Still TODO:
- Consider not adding the content filtering to the `comment_text` filter. Comments do not support images by default. If a plugin wants to enable images in comments, it can easily add that filter.

**Comment by azaozz** [comment]
In https://github.com/WordPress/wp-lazy-loading/pull/23/commits/060288bff08f0432bd43218cf49ba615644d6c54 also removed passing the `attachment_id` to `wp_image_tag_add_loading_attr()`. Reasons:
- It is not needed for the loading attr.
- There is a really annoying mismatch of the found attachment IDs in post_content when the WP posts have been exported from one site and then imported in another (it is tested for/avoided when adding srcset and sizes). Passing the ID here as an optional param gives the wrong impression that it can always be relied on...

File: wp-lazy-loading.php [hunk 1]
```diff
@@ -122,15 +122,12 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
...
- * * add 'srcset' and 'sizes' attributes to 'img' tags
- * * add 'loading' attributes to 'img' tags
+ * This function adds `srcset`, `sizes`, and `loading` attributes to `img` HTML tags.
  *
  * @since (TBD)
  *
- * @see wp_image_add_srcset_and_sizes()
- * @see wp_image_add_loading()
+ * @see wp_image_tag_add_loading_attr()
+ * @see wp_image_tag_add_srcset_and_sizes_attr()
```
Comment (felixarntz) [inline wp-lazy-loading.php]: I wonder whether we should name these `wp_img_tag_*`, after all that's the actual tag.

File: wp-lazy-loading.php [hunk 2]
```diff
@@ -141,19 +138,21 @@ function wp_filter_content_tags( $content, $context = null ) {
 		$context = current_filter();
 	}
 
-	$add_srcset_sizes = 'the_content' === $context;
 	$add_loading_attr = wp_lazy_loading_enabled( 'img', $context );
 
-	if ( false === strpos( $content, '<img' ) || ( ! $add_srcset_sizes && ! $add_loading_attr ) ) {
+	if ( false === strpos( $content, '<img' ) ) {
```
Comment (felixarntz) [inline wp-lazy-loading.php:143]: I don't have a strong opinion here, but by removing the `$add_srcset_sizes` condition we actually change the behavior of where `srcset` and `sizes` are added (currently only supported for `the_content`). This is kind of an unintended side effect (this is only about `loading`), therefore I would rather have us keep the current behavior to only do that in `the_content`. Changing that is for another discussion IMO.

If we keep it, then having this condition can avoid unnecessary execution of the heavier logic below.

File: wp-lazy-loading.php [hunk 3]
```diff
@@ -118,57 +120,153 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
...
+
+	// List of the unique `img` tags found in $content.
+	$images = array();
 
-				if ( $value ) {
-					if ( ! in_array( $value, array( 'lazy', 'eager' ), true ) ) {
-						$value = 'lazy';
-					}
+	// List of the rest of the images, either local but without attachment_id, or remote.
+	$other_images = array();
```
Comment (felixarntz) [inline wp-lazy-loading.php]: This is an unused remainder and should be removed :)

File: wp-lazy-loading.php [hunk 4]
```diff
@@ -118,57 +120,153 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
...
+
+		// Add 'loading' attribute if applicable.
+		if ( $add_loading_attr && false === strpos( $filtered_image, ' loading=' ) ) {
+			$filtered_image = wp_image_tag_add_loading_attr( $filtered_image, $context );
+		}
+
+		if ( $filtered_image !== $image ) {
+			$content = str_replace( $image, $filtered_image, $content );
+		}
+	}
```
Comment (felixarntz) [inline wp-lazy-loading.php:198]: Nice, I really like the single loop here used here.

File: wp-lazy-loading.php [hunk 5]
```diff
@@ -118,57 +120,153 @@ function wp_lazy_loading_enabled( $tag_name, $context ) {
...
+	 * Returning anything else than `true` will not add the attributes.
+	 *
+	 * @since (TBD)
+	 *
+	 * @param bool   $value         The filtered value, defaults to `true`.
+	 * @param string $image         The HTML `img` tag where the attribute should be added.
+	 * @param string $context       Additional context about how the function was called or where the img tag is.
+	 * @param int    $attachment_id The image attachment ID.
+	 */
+	$add = apply_filters( 'wp_image_tag_add_srcset_and_sizes_attr', true, $image, $context, $attachment_id );
```
Comment (felixarntz) [inline wp-lazy-loading.php]: I'm not sure whether we should introduce a filter like this. I get the idea and am generally not opposed, but this is an unrelated change and would raise further discussion points that don't actually help what we want to achieve here.

I'd say let's simplify this function to only be the part within the below if clause.

**Review by felixarntz — CHANGES_REQUESTED** [review]
@azaozz This is looking great. I have a few follow-up comments. Let's discuss and address these, but afterwards I think this is good to be merged. At minimum, one unnecessary piece of code needs to be changed here :)

**Comment by felixarntz** [comment]
@azaozz 
> Still TODO:
> 
> * Consider not adding the content filtering to the `comment_text` filter. Comments do not support images by default. If a plugin wants to enable images in comments, it can easily add that filter.

That sounds good to me, let's remove it.

File: wp-lazy-loading.php (hunk 1 above)
Comment (azaozz) [reply wp-lazy-loading.php]: Sounds good.

File: wp-lazy-loading.php (hunk 2 above)
Comment (azaozz) [reply wp-lazy-loading.php:143]: Yeah, went back and forth on this a few times. If we keep it, it makes `wp_filter_content_tags()` a bit harder to use for a "random" HTML string/content. It is possible to use `wp_make_content_images_responsive()` like that.

If we remove `$add_srcset_sizes`, by default this will run on `the_content`, `the_excerpt`, and `widget_text_content` (after removing `comment_text`). Thinking `wp_make_content_images_responsive()` should have been running on both `the_excerpt`, and `widget_text_content` from when it was added. It makes sense to add srcset, sizes, and loading attributes to images there. 

> ...can avoid unnecessary execution of the heavier logic below.

It will (most likely) be running the `preg_match_all()` anyway. Then the difference would be `preg_match( '/wp-image-([0-9]+)/i', $image, $class_id )` on the img tags which should be really fast as these would usually be few short strings.

File: wp-lazy-loading.php (hunk 3 above)
Comment (azaozz) [reply wp-lazy-loading.php]: Oops, missed it :)

File: wp-lazy-loading.php (hunk 5 above)
Comment (azaozz) [reply wp-lazy-loading.php]: Yeah, still on the fence about it too.

Having a filter somewhat matches `wp_image_tag_add_loading_attr()` (and eventually `wp_iframe_tag_add_loading_attr()`). It also lets plugin disable srcset and sizes depending on `$context`, which might be useful in some edge cases but not strictly necessary.

Not having a filter makes introducing `wp_image_tag_add_srcset_and_sizes_attr()` somewhat questionable as the same code may as well be inside the loop where that function is called...

Perhaps can keep it for now and revisit when making the core patch? Then there will likely be more people looking and helping :)

File: wp-lazy-loading.php (hunk 2 above)
Comment (felixarntz) [reply wp-lazy-loading.php:143]: I'm okay with this change (also supporting the other filters for `srcset` and `sizes`), but we need to clarify that when talking about the core patch (and if we keep it also in the dev note).
From a technical standpoint it makes sense, I just wonder whether it will result in unrelated discussion and overcomplicate these efforts. But we can change still change it back if undesirable per core patch discussion.

File: wp-lazy-loading.php (hunk 5 above)
Comment (felixarntz) [reply wp-lazy-loading.php]: Actually revisiting this, we should probably have this filter if we make the other change from above (supporting more than `the_content` here). With this filter, we'll allow developers to opt out for other contexts than `the_content` (not sure anybody will wanna do that, but at least it allows someone to get back to the previous behavior if they have some random reason to do so).

**Review by felixarntz — APPROVED** [review]

**Review by felixarntz — APPROVED** [review]

File: wp-lazy-loading.php (hunk 2 above)
Comment (azaozz) [reply wp-lazy-loading.php:143]: > we need to clarify that when talking about the core patch
> ...
> I just wonder whether it will result in unrelated discussion

Yeah, might happen. Thinking to separate this into another trac ticket. Just want to avoid hard-coding it there.
